File: src/main/java/miragefairy2019/mod/lib/multi/ItemMulti.java
package miragefairy2019.mod.lib.multi;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

import mirrg.boron.util.struct.Tuple;
import mirrg.boron.util.suppliterator.ISuppliterator;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.NonNullList;

public class ItemMulti<V extends ItemVariant> extends Item
{

	public ItemMulti()
	{
		this.setHasSubtypes(true);
		this.setMaxDamage(0);
	}

	//

	private Map<Integer, V> map = new HashMap<>();

	public void registerVariant(int metadata, V variant)
	{
		if (map.containsKey(metadata)) throw new IllegalArgumentException("Illegal metadata: " + metadata);
		map.put(metadata, variant);
	}

	public Optional<V> getVariant(ItemStack stack)
	{
		return getVariant(stack.getMetadata());
	}

	public Optional<V> getVariant(int metadata)
	{
		return Optional.ofNullable(map.get(metadata));
	}

	public Iterable<Tuple<Integer, V>> getVariants()
	{
		return ISuppliterator.ofIterable(map.entrySet())
			.map(e -> Tuple.of(e.getKey(), e.getValue()));
	}

	//

	@Override
	public String getUnlocalizedName(ItemStack stack)
	{
		return getVariant(stack).map(v -> v.getUnlocalizedName()).orElse("item.null");
	}

	@Override
	public void getSubItems(CreativeTabs tab, NonNullList<ItemStack> items)
	{
		if (!isInCreativeTab(tab)) return;
		map.keySet().forEach(metadata -> {
			items.add(new ItemStack(this, 1, metadata));
		});
	}

}

File: src/main/java/miragefairy2019/mod/lib/ItemMulti.kt
package miragefairy2019.mod.lib

import miragefairy2019.libkt.canTranslate
import miragefairy2019.libkt.createItemStack
import miragefairy2019.libkt.translateToLocal
import net.minecraft.client.renderer.block.model.ModelResourceLocation
import net.minecraft.client.util.ITooltipFlag
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.util.NonNullList
import net.minecraft.util.ResourceLocation
import net.minecraft.world.World
import net.minecraftforge.client.model.ModelLoader
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly

open class ItemMulti<V : ItemVariant> : Item() {
    init {
        setHasSubtypes(true)
        maxDamage = 0
    }

    //

    private val map = mutableMapOf<Int, V>()

    fun registerVariant(metadata: Int, variant: V) {
        require(metadata !in map) { "Illegal metadata: $metadata" }
        map[metadata] = variant
        variant.item = this
        variant.metadata = metadata
    }

    fun getVariant(itemStack: ItemStack) = getVariant(itemStack.metadata)

    fun getVariant(metadata: Int) = map[metadata]

    val variants get() = map.values

    //

    override fun getSubItems(tab: CreativeTabs, items: NonNullList<ItemStack>) {
        if (!isInCreativeTab(tab)) return
        map.forEach { (metadata, variant) ->
            if (variant.canSeeOnCreativeTab) items += createItemStack(1, metadata)
        }
    }
}

open class ItemVariant {
    @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
    private lateinit var iMetadata: Integer
    var metadata
        get() = iMetadata.toInt()
        set(it) {
            iMetadata = Integer(it)
        }
    lateinit var item: Item

    fun createItemStack(amount: Int = 1) = item.createItemStack(amount, metadata)
    var canSeeOnCreativeTab = true
}


// Material

open class ItemMultiMaterial<V : ItemVariantMaterial> : ItemMulti<V>() {
    override fun getUnlocalizedName(itemStack: ItemStack): String {
        val variant = getVariant(itemStack) ?: return "item.null"
        return "item.${variant.unlocalizedName}"
    }

    // TODO 子クラスに移動
    @SideOnly(Side.CLIENT)
    override fun addInformation(itemStack: ItemStack, world: World?, tooltip: MutableList<String>, flag: ITooltipFlag) {
        getVariant(itemStack) ?: return

        // ポエム
        val key = "${getUnlocalizedName(itemStack)}.poem"
        if (canTranslate(key)) {
            val string = translateToLocal(key)
            if (string.isNotEmpty()) tooltip += string
        }

    }
}

@SideOnly(Side.CLIENT)
fun <V : ItemVariantMaterial> ItemMultiMaterial<V>.setCustomModelResourceLocations() = variants.forEach { variant ->
    ModelLoader.setCustomModelResourceLocation(
        this,
        variant.metadata,
        ModelResourceLocation(ResourceLocation(registryName!!.resourceDomain, variant.registryName), "normal")
    )
}

open class ItemVariantMaterial(val registryName: String, val unlocalizedName: String) : ItemVariant()
